packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,1323 @@
+	          femv_colors(curr_lid) = 0;
+		  recoloring_size++;
+	          break;
+	        } else if(rand_dev(vid_d2) < rand_dev(curr_lid)){
+	          femv_colors(vid_d2) = 0;
+		  recoloring_size++;
+	        } else {
+	          if(gid_dev(curr_lid) >= gid_dev(vid_d2)){
+	            found = true;
+	            femv_colors(curr_lid) = 0;
+		    recoloring_size++;
+	            break;
+	          } else {
+	            femv_colors(vid_d2) = 0;
+		    recoloring_size++;
+	          }
+ 	        }
+	      }
+	    }
+       	    if(found) break;
+          }
+        },recoloringSize(0));
+        Kokkos::fence();
+	verts_to_send_size(0) = 0;
+	verts_to_recolor_size(0) = 0;
+	Kokkos::parallel_for(femv_colors.size(), KOKKOS_LAMBDA (const size_t& i){
+	  if(femv_colors(i) == 0){
+	    if(i < n_local){
+	      verts_to_send_atomic(verts_to_send_size_atomic(0)++) = i;
+	    }
+	    verts_to_recolor_atomic(verts_to_recolor_size_atomic(0)++) = i;
+	  }		
+	});
+        Kokkos::fence();
+        conflictDetectionPerRound[distributedRounds] = timer() - detection_temp;
+        conflict_detection += conflictDetectionPerRound[distributedRounds];
+        compPerRound[distributedRounds] += conflictDetectionPerRound[distributedRounds];
+        totalPerRound[distributedRounds] += conflictDetectionPerRound[distributedRounds];
+        comp_time += conflictDetectionPerRound[distributedRounds];
+
+        distributedRounds++;
+        int localDone = recoloringSize(0);
+        int globalDone = 0;
+        Teuchos::reduceAll<int,int>(*comm, Teuchos::REDUCE_SUM, 1, &localDone, &globalDone);
+        done = !globalDone;
+      
+      }//end coloring
+
+      if(recoloringSize(0) > 0 || !done){
+        //Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<memory_space>();
+	//Kokkos::View<int*, device_type> femv_colors = subview(femvColors, Kokkos::ALL,0);
+	colors_host = Kokkos::create_mirror_view(femv_colors);
+	deep_copy(colors_host, femv_colors);
+	ghost_colors_host = Kokkos::create_mirror_view(ghost_colors);
+	deep_copy(ghost_colors_host, ghost_colors);
+	verts_to_recolor_host = Kokkos::create_mirror_view(verts_to_recolor_view);
+	deep_copy(verts_to_recolor_host, verts_to_recolor_view);
+	verts_to_send_host = Kokkos::create_mirror_view(verts_to_send_view);
+	deep_copy(verts_to_send_host, verts_to_send_view);
+      }
+
+      while(recoloringSize(0) > 0 || !done){
+        //Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<memory_space>();
+	//Kokkos::View<int*, device_type> femv_colors = subview(femvColors, Kokkos::ALL, 0);
+	if(distributedRounds < 100){
+	  vertsPerRound[distributedRounds] = recoloringSize(0);
+	}
+	if(verbose) std::cout<<comm->getRank()<<": starting to recolor, serial\n";
+        comm->barrier();
+	double recolor_temp = timer();
+	if(verts_to_recolor_size(0) > 0){
+	  this->colorInterior<Kokkos::DefaultHostExecutionSpace,
+		              memory_space,
```

Comment:
Templated memory space should be Host or HostMirror space.  With UVM, this code works, but without UVM, memory_space is GPU memory.